node/src/Pos/Explorer/DB.hs

```diff
@@ -104,10 +133,116 @@ prepareExplorerDB = do
+
+            convertToPaged :: (Epoch, [HeaderHash]) -> [((Epoch, Page), [HeaderHash])]
+            convertToPaged (epoch, headerHashes) =
+                convertHHsToEpochPages <$> convertHHsToPages
+              where
+                convertHHsToEpochPages
+                    :: (Page, [HeaderHash])
+                    -> ((Epoch, Page), [HeaderHash])
+                convertHHsToEpochPages (page, headerHashes') =
+                    ((epoch, page), headerHashes')
+
+                -- | Take that huge chunk of @HeaderHash@es and page it.
+                convertHHsToPages :: [(Page, [HeaderHash])]
+                convertHHsToPages = zip [1..] $ splitEvery pageSize headerHashes
+                  where
+                    -- | TODO (ks): It's getting repeated all over the place.
+                    pageSize :: Int
+                    pageSize = 10
+
+                    splitEvery :: Int -> [a] -> [[a]]
+                    splitEvery _ [] = []
+                    splitEvery n xs = as : splitEvery n bs
+                      where
+                        (as,bs) = splitAt n xs
+
+        -- | Finally, we persist the map with the new format.
+        epochPagesExplorerOpConduit
+            :: (Monad m)
+            => Conduit (Map EpochPagedBlocksKey [HeaderHash]) m [ExplorerOp]
+        epochPagesExplorerOpConduit = CL.map persistEpochBlocks
+          where
+            -- | Persist atomically, all the operations together.
+            persistEpochBlocks :: Map EpochPagedBlocksKey [HeaderHash] -> [ExplorerOp]
+            persistEpochBlocks mapEpochPagedHHs =
+                persistEpochPageBlocks ++ persistMaxPageNumbers
+              where
+                -- | Persist @Epoch@ @Page@ blocks.
+                persistEpochPageBlocks :: [ExplorerOp]
+                persistEpochPageBlocks = putKeyBlocks <$> M.toList mapEpochPagedHHs
+                  where
+                    putKeyBlocks
+                        :: (EpochPagedBlocksKey, [HeaderHash])
+                        -> ExplorerOp
+                    putKeyBlocks keyBlocks = PutEpochBlocks epoch page blocks
+                      where
+                        key           = keyBlocks ^. _1
+                        epoch         = key ^. _1
+                        page          = key ^. _2
+
+                        blocks        = keyBlocks ^. _2
+
+                -- | Persist @Epoch@ max @Page@.
+                persistMaxPageNumbers :: [ExplorerOp]
+                persistMaxPageNumbers =
+                    [ PutEpochPages epoch maxPageNumber | (epoch, maxPageNumber) <- emp]
+                  where
+                    emp :: [EpochPagedBlocksKey]
+                    emp = findEpochMaxPages mapEpochPagedHHs
+
+        -- | Finally, just persist all the operations atomically.
+        persistExplorerOpSink :: (MonadDB m) => Sink ([ExplorerOp]) m ()
+        persistExplorerOpSink = CL.mapM_ writeBatchGState
```

Comment:
This is not atomic. Essentially, you are persisting each of the `Op`s atomically, but not all of them at once.

Reply:
Isn't `Batch` used for "all or nothing" situations? I'll take a look at this in more detail, thanks.

Reply:
It's atomic.

Reply:
Ah, I think I see what you mean. Will fix.